jwst/associations/lib/rules_level2_base.py

```diff
@@ -583,138 +586,122 @@ def _merge_asns(asns):
-            asn
-            for asn in merged.values()
-        ]
+        merged_asns = list(merged.values())
         return merged_asns
 
 
 # -----------------
 # Basic constraints
 # -----------------
 class Constraint_Base(Constraint):
-    """Select on program and instrument"""
+    """Select on program and instrument."""
 
     def __init__(self):
-        super(Constraint_Base, self).__init__([
-            DMSAttrConstraint(
-                name='program',
-                sources=['program']
-            ),
-            DMSAttrConstraint(
-                name='is_tso',
-                sources=['tsovisit'],
-                required=False,
-                force_unique=True,
-            )
-        ])
+        super(Constraint_Base, self).__init__(
+            [
+                DMSAttrConstraint(name="program", sources=["program"]),
+                DMSAttrConstraint(
+                    name="is_tso",
+                    sources=["tsovisit"],
+                    required=False,
+                    force_unique=True,
+                ),
+            ]
+        )
 
 
 class Constraint_Background(DMSAttrConstraint):
-    """Select backgrounds"""
+    """Select backgrounds."""
 
     def __init__(self):
         super(Constraint_Background, self).__init__(
-            sources=['bkgdtarg'],
+            sources=["bkgdtarg"],
             force_unique=False,
-            name='background',
+            name="background",
             force_reprocess=ListCategory.EXISTING,
             only_on_match=True,
         )
 
 
 class Constraint_ExtCal(Constraint):
-    """Remove any nis_extcals from the associations, they
-       are NOT to receive level-2b or level-3 processing!"""
+    """
+    Remove any nis_extcals from the associations.
+
+    They are NOT to receive level-2b or level-3 processing.
+    """
 
     def __init__(self):
         super(Constraint_ExtCal, self).__init__(
-            [
-                DMSAttrConstraint(
-                    name='exp_type',
-                    sources=['exp_type'],
-                    value='nis_extcal'
-                )
-            ],
-            reduce=Constraint.notany
+            [DMSAttrConstraint(name="exp_type", sources=["exp_type"], value="nis_extcal")],
+            reduce=Constraint.notany,
         )
 
 
 class Constraint_Imprint(Constraint):
-    """Select on imprint exposures"""
+    """Select on imprint exposures."""
 
     def __init__(self):
         super(Constraint_Imprint, self).__init__(
-            [
-                DMSAttrConstraint(
-                    name='imprint',
-                    sources=['is_imprt']
-                )
-            ],
+            [DMSAttrConstraint(name="imprint", sources=["is_imprt"])],
             reprocess_on_match=True,
             work_over=ListCategory.EXISTING,
         )
 
 
 class Constraint_Imprint_Special(Constraint):
-    """Select on imprint exposures"""
+    """Select on imprint exposures."""
```

Comment:
Can you add something here to differentiate "special" from the regular class above?